cli/internal/core/engine.go

```diff
@@ -257,3 +268,77 @@ func (e *Engine) AddDep(fromTaskID string, toTaskID string) error {
 
 	return nil
 }
+
+// ValidatePersistentDependencies checks if any task dependsOn persistent tasks and throws
+// an error if that task is actually implemented
+func (e *Engine) ValidatePersistentDependencies(graph *graph.CompleteGraph) error {
+	var validationError error
+
+	// Adding in a lock because otherwise walking the graph can introduce a data race
+	// (reproducible with `go test -race`)
+	var sema = util.NewSemaphore(1)
+
+	errs := e.TaskGraph.Walk(func(v dag.Vertex) error {
+		vertexName := dag.VertexName(v) // vertexName is a taskID
+
+		// No need to check the root node if that's where we are.
+		if strings.Contains(vertexName, ROOT_NODE_NAME) {
+			return nil
+		}
+
+		// Aquire a lock, because otherwise walking this group can cause a race condition
+		// writing to the same validationError var defined outside the Walk(). This shows
+		// up when running tests with the `-race` flag.
+		sema.Acquire()
+		defer sema.Release()
+
+		currentPackageName, currentTaskName := util.GetPackageTaskFromId(vertexName)
+
+		// For each "downEdge" (i.e. each task that _this_ task dependsOn)
+		// check if the downEdge is a Persistent task, and if it actually has the script implemented
+		// in that package's package.json
+		for dep := range e.TaskGraph.DownEdges(vertexName) {
```

Comment:
I will validate this with a test in a couple hours, but I'm 99% sure that `package-1#dev` would be one of the nodes while walking the graph, and it would check its dependencies on its own.

Reply:
For the test of this I want to make sure we have good coverage of skipped-layer stuff, skipped tasks, branching into different task names, and named non-topo deps. e.g.:

"safe" ^dev:
```
dev dependsOn [^dev]
packages a, b, c; c => b => a

but only _one_ of them actually has a `dev` script.
```

skip level:
```
c dependsOn [^dev]
a#dev => b(nonexistent devscript) => c#dev
```

Topo with weird sidecar:
```
packages a, b, c; c => b => a
build dependsOn [^build]
a#build dependsOn [z#dev]
```

Some different task dep switching out of topo deps:
```
non-topo deps (no dependency relationship in package.json):
z#dev persistent
c#lint dependsOn [b#lint]
b#lint dependsOn [a#lint]
a#lint dependsOn [z#dev]
```

etc.

Reply:
(I updated your comment with numbers, so I could respond without noisy quoting)

1. will do
2. don't understand this one. In `c dependsOn [^dev]`, is c a task name or package name?
3. i can add this, but I'd be very surprised if it makes a material difference to the algo
4. same as 3

more tests incoming